### PR TITLE
Remove stale non-Standard-GQA gate for Nemotron BCG

### PR DESCRIPTION
**Pull request by b8zhong**
It's actually fully supported after https://github.com/sgl-project/sglang/pull/21436 already. I think this guard was added back as a mistake. Otherwise, the prefill TTFT is really dominated by host overhead

For example, Qwen 3.5 another GQA and linear attention model fully supports BCG. https://github.com/sgl-project/sglang/pull/30620

```
python3 -m sglang.launch_server \
  --model-path nvidia/NVIDIA-Nemotron-3-Ultra-550B-A55B-NVFP4 \
  --trust-remote-code \
  --tp 4 \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 5 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 6 \
  --mamba-backend flashinfer \
  --mamba-ssm-dtype float32 \
  --reasoning-parser nemotron_3 \
  --tool-call-parser qwen3_coder \
  --mamba-scheduler-strategy extra_buffer \
  --attention-backend trtllm_mha --chunked-prefill-size 8192
  ```
  
  ```
  [2026-07-14 15:46:38 TP0] Capture target prefill CUDA graph begin. backend=breakable, num_tokens=[4, 8, 12, 16, 20, 24, 28, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256, 288, 320, 352, 384, 416, 448, 480, 512, 576, 640, 704, 768, 832, 896, 960, 1024, 1280, 1536, 1792, 2048, 2304, 2560, 2816, 3072, 3328, 3584, 3840, 4096, 4608, 5120, 5632, 6144, 6656, 7168, 7680, 8192], avail mem=15.39 GB
[2026-07-14 15:46:38 TP3] FlashInfer autotune completed.
[2026-07-14 15:46:38 TP3] Capture target prefill CUDA graph begin. backend=breakable, num_tokens=[4, 8, 12, 16, 20, 24, 28, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256, 288, 320, 352, 384, 416, 448, 480, 512, 576, 640, 704, 768, 832, 896, 960, 1024, 1280, 1536, 1792, 2048, 2304, 2560, 2816, 3072, 3328, 3584, 3840, 4096, 4608, 5120, 5632, 6144, 6656, 7168, 7680, 8192], avail mem=15.64 GB
[2026-07-14 15:46:38 TP1] FlashInfer autotune completed.
[2026-07-14 15:46:38 TP1] Capture target prefill CUDA graph begin. backend=breakable, num_tokens=[4, 8, 12, 16, 20, 24, 28, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256, 288, 320, 352, 384, 416, 448, 480, 512, 576, 640, 704, 768, 832, 896, 960, 1024, 1280, 1536, 1792, 2048, 2304, 2560, 2816, 3072, 3328, 3584, 3840, 4096, 4608, 5120, 5632, 6144, 6656, 7168, 7680, 8192], avail mem=15.33 GB
[2026-07-14 15:46:38 TP2] FlashInfer autotune completed.
[2026-07-14 15:46:38 TP2] Capture target prefill CUDA graph begin. backend=breakable, num_tokens=[4, 8, 12, 16, 20, 24, 28, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256, 288, 320, 352, 384, 416, 448, 480, 512, 576, 640, 704, 768, 832, 896, 960, 1024, 1280, 1536, 1792, 2048, 2304, 2560, 2816, 3072, 3328, 3584, 3840, 4096, 4608, 5120, 5632, 6144, 6656, 7168, 7680, 8192], avail mem=15.33 GB
Capturing num tokens (num_tokens=4 avail_mem=9.86 GB): 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 58/58 [00:22<00:00,  2.62it/s]
[2026-07-14 15:47:02 TP3] Capture target prefill CUDA graph end. elapsed=23.62 s, mem usage=3.21 GB, avail mem=12.43 GB.
  ```
  
  Before
  ```
  python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 128 --platinum
Loading GSM8K Platinum dataset from HuggingFace...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1209/1209 [00:47<00:00, 25.38it/s]
Accuracy: 0.959
Invalid: 0.000
Latency: 47.634 s
Output throughput: 2814.054 token/s
```

After
```
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 128 --platinum
Loading GSM8K Platinum dataset from HuggingFace...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1209/1209 [00:43<00:00, 28.01it/s]
Accuracy: 0.959
Invalid: 0.000
Latency: 43.165 s
Output throughput: 3084.422 token/s
```

---
### CI States

Latest PR Test (Base): :x: [Run #34828382059](https://github.com/sgl-project/sglang/actions/runs/34828382059)
Latest PR Test (Extra): :x: [Run #34828381572](https://github.com/sgl-project/sglang/actions/runs/34828381572)
Latest PR Test (AMD ROCm 10): :x: [Run #34828381808](https://github.com/sgl-project/sglang/actions/runs/34828381808)

### CODE REVIEW

File: python/sglang/srt/models/nemotron_h_mtp.py [hunk 1]
```diff
@@ -288,13 +288,14 @@ def get_input_embeddings(self, input_ids: torch.Tensor) -> torch.Tensor:
     def forward(
         self,
         input_ids: torch.Tensor,
-        hidden_states: torch.Tensor,
```
Comment (b8zhong) [inline python/sglang/srt/models/nemotron_h_mtp.py]: This is consistent with the pattern of DSV2 MTP. Otherwise, will encounter an error

**Comment by ksh108405** [comment]
We had some accuracy regression in production on NVIDIA-Nemotron-3-Ultra-550B (NVFP4, `nemotron_h` hybrid, 8×H100, TP8) that a single-shot benchmark won't surface.

## What we observed

With the prefill CUDA graph (BCG) enabled, GSM8K accuracy degrades in proportion to radix/mamba
cache hit depth.

```
prefill CUDA graph ON,  repeated warm GSM8K (200q ×N, same server):
  run1 (cold)  0.965
  run2         0.900
  run3         0.920
  run4         0.915       ← latency drops 21s→15s across runs = cache hits deepening
prefill CUDA graph OFF: 0.965 stable, restored immediately on disable
```

A single benchmark pass reads ~0.95–0.96 (the cold number) and looks healthy, and the regression only
appears once the mamba state written by a graph-executed prefill is **read back from cache** on a
later request.

## Controlled isolation (fresh restart per arm, GSM8K 200q ×3, identical flags)

| Arm | graph | mamba-state cache | run1 → run3 |
|-----|-------|-------------------|-------------|
| R | ON | int8-mamba-checkpoint ON | 0.950 → **0.905 → 0.905** (sticky −6pp) |
| A | ON | int8 OFF, plain radix | 0.960 → 0.935 → 0.950 (weak, −1~2pp) |
| control | OFF | int8 ON | 0.955–0.970 stable |

I think the dominant effect is the interaction between graph-executed prefill and cached mamba state
(worst with the int8 checkpoint pool).

## Ask

If `main` removes the BCG gate without a companion fix for the mamba-state a graphed prefill writes into radix/int8 cache, this reintroduces a cache-depth-dependent accuracy regression that CI's single-shot GSM8K won't catch.
Could we either (a) gate BCG off when mamba-state caching (radix v2 / int8-mamba-checkpoint) is active, or (b) add a warm/repeated-GSM8K arm to the validation for hybrid models?

Happy to share the full harness (repeated-warm GSM8K driver + per-arm logs) if useful. Environment:
sglang v0.5.15, `--speculative-algorithm EAGLE` `--speculative-num-steps=7` `--mamba-radix-cache-strategy extra_buffer` `--enable-int8-mamba-checkpoint`, fa3 + marlin MoE on SM90.

File: python/sglang/srt/model_executor/model_runner_components/layer_setup.py [hunk 2]
```diff
@@ -23,7 +24,10 @@ def compute_attention_and_moe_layers(layer_model: Any) -> AttentionAndMoeLayers:
     moe_fusions: list[Any] = []
     dsa_indexers: list[Any] = []
     mha_companion_layers: list[Any] = []
-    for layer in layer_model.layers:
+    layers = layer_model.layers
+    if isinstance(layers, nn.ModuleDict):
```
Comment (b8zhong) [inline python/sglang/srt/model_executor/model_runner_components/layer_setup.py:28]: Note: Nemotron builds as ModelDict.

**Comment by b8zhong** [comment]
@ksh108405 Hi, thanks for this information. Could you share: the server arg and evaluation commands?

**Comment by ksh108405** [comment]
@b8zhong Here is the scripts:
Model: `nvidia/NVIDIA-Nemotron-3-Ultra-550B-A55B-NVFP4`, 8×H100 (SM90),
sglang v0.5.15 + #29931 

### Server args

Baseline (prefill CUDA graph off):

```bash
python3 -m sglang.launch_server \
  --model-path <MODEL> --trust-remote-code \
  --host 0.0.0.0 --port 30000 \
  --tp-size 8 \
  --context-length 262144 \
  --kv-cache-dtype fp8_e4m3 \
  --mamba-radix-cache-strategy extra_buffer \
  --mamba-backend flashinfer --mamba-ssm-dtype float16 \
  --enable-mamba-cache-stochastic-rounding \
  --speculative-algorithm EAGLE --speculative-num-steps 7 \
  --speculative-eagle-topk 1 --speculative-num-draft-tokens 8 \
  --speculative-draft-model-path <MODEL> \
  --max-running-requests 48 \
  --cuda-graph-max-bs-decode 64 \
  --mem-fraction-static 0.85 \
  --chunked-prefill-size 8192 \
  --reasoning-parser nemotron_3 --tool-call-parser qwen3_coder \
  --mamba-cache-philox-rounds 5 \
  --enable-int8-mamba-checkpoint \
  --disable-prefill-cuda-graph
```

The three arms differ by only two flags:

| Arm | prefill CUDA graph | mamba-state cache | flag delta vs baseline |
|-----|--------------------|-------------------|------------------------|
| **control** | OFF | int8 on | (baseline as above) |
| **A** | ON | plain radix (int8 off) | **remove** `--disable-prefill-cuda-graph`, **add** `--cuda-graph-max-bs-prefill 8192`, **remove** `--enable-int8-mamba-checkpoint` |
| **R** | ON | int8-mamba-checkpoint | **remove** `--disable-prefill-cuda-graph`, **add** `--cuda-graph-max-bs-prefill 8192` (keep `--enable-int8-mamba-checkpoint`) |

### Evaluation commands

sglang's built-in GSM8K, run three times back-to-back against the same server:

```bash
for run in 1 2 3; do
  echo "=== run $run ==="
  python3 -m sglang.test.few_shot_gsm8k --num-questions 200 --port 30000 --parallel 64
done
```

we should look for accuracy decays across the three runs while latency drops. On our runs:

```
Arm R (graph ON + int8):   0.950 -> 0.905 -> 0.905   (latency 25.6s -> 14.8s)
Arm A (graph ON, no int8): 0.960 -> 0.935 -> 0.950
control (graph OFF):       0.955 - 0.970 stable, restored immediately on disabling the graph
```

Between arms we do a fresh restart (the cache must be cold for run 1).

**Comment by b8zhong** [comment]
@ksh108405 I've verified the performance drop, even without int8 on B300. The result definitely seems to be abnormal, especially the fact that the invalid number goes from 0 to 0.015 at the third round, as well, flush_cache can recover the accuracy again.

```
for run in 1 2 3; do
  echo "=== run $run ==="
  python3 -m sglang.test.few_shot_gsm8k \
    --num-questions 200 \
    --port 30000 \
    --parallel 64
done
=== run 1 ===
/sgl-workspace/sglang/python/sglang/test/few_shot_gsm8k.py:166: DeprecationWarning: sglang.test.few_shot_gsm8k is deprecated. Use sglang.test.run_eval with eval_name='gsm8k' instead.
  run_eval(args)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:09<00:00, 20.79it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 10.517 s
Output throughput: 1978.326 token/s
=== run 2 ===
/sgl-workspace/sglang/python/sglang/test/few_shot_gsm8k.py:166: DeprecationWarning: sglang.test.few_shot_gsm8k is deprecated. Use sglang.test.run_eval with eval_name='gsm8k' instead.
  run_eval(args)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:06<00:00, 30.17it/s]
Accuracy: 0.895
Invalid: 0.010
Latency: 6.708 s
Output throughput: 2805.489 token/s
=== run 3 ===
/sgl-workspace/sglang/python/sglang/test/few_shot_gsm8k.py:166: DeprecationWarning: sglang.test.few_shot_gsm8k is deprecated. Use sglang.test.run_eval with eval_name='gsm8k' instead.
  run_eval(args)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:06<00:00, 30.98it/s]
Accuracy: 0.895
Invalid: 0.015
Latency: 6.528 s
Output throughput: 2919.675 token/s
```

**Comment by b8zhong** [comment]
Hi @ksh108405 , after some testing, it appears this only happens as a result of BCG and speculative decoding running together. Do you have any insights about that? I'm also continuing to debug this issue. Please message me on Slack about this (Brayden Zhong, the one with the panda) (since my Github inbox is really chaos)

File: python/sglang/kernels/ops/mamba/triton_ops/ssd_chunk_scan.py [hunk 3]
```diff
@@ -272,7 +272,7 @@ def _chunk_scan_fwd_kernel(
             else:
                 # - if there is initstates, we will rely on prev_states, no zeroing
                 #   required.
-                scale_m = tl.exp(dA_cs_m - dA_cs_m_boundary)
```
Comment (b8zhong) [inline python/sglang/kernels/ops/mamba/triton_ops/ssd_chunk_scan.py:271]: Numerical stability fix from upstream FLA: https://github.com/state-spaces/mamba/pull/713

File: python/sglang/kernels/ops/mamba/triton_ops/ssd_chunk_state.py [hunk 4]
```diff
@@ -229,10 +229,12 @@ def _chunk_state_fwd_kernel(
             tl.float32
         )
         if not HAS_SEQ_IDX:
-            scale = tl.exp(dA_cs_last - dA_cs_k) * dt_k
```
Comment (b8zhong) [inline python/sglang/kernels/ops/mamba/triton_ops/ssd_chunk_state.py:232]: Same as https://github.com/state-spaces/mamba/pull/713

File: python/sglang/srt/layers/attention/mamba/mamba2_metadata.py [hunk 5]
```diff
@@ -237,7 +237,7 @@ def prepare_mixed(
         batch_size = getattr(forward_batch, "_original_batch_size", None)
         if batch_size is None:
             batch_size = len(forward_batch.seq_lens)
-        num_decodes = batch_size - num_prefills
```
Comment (b8zhong) [inline python/sglang/srt/layers/attention/mamba/mamba2_metadata.py]: When running with DP attention and BS = 1, the num prefill will be 1, but the batch size with decode will be 0, causing a bug when the num decodes is -1.

**Comment by b8zhong** [comment]
Hi @ksh108405 , when setting the --mamba-track-interval 128 alongside this PR, is it fully fixed for you?

**Comment by ksh108405** [comment]
Hi @b8zhong, I tracked it down, and found that this PR is not the cause. sorry for pointing you at it.

## Actual root cause

`PrefillCudaGraphRunner._is_mamba_track_enabled()`:

```python
return (
    self.model_runner.server_args.enable_mamba_extra_buffer()
    and not self.model_runner.server_args.disable_radix_cache
    and self.model_runner.spec_algorithm.is_none()   # <-- here
)
```

With speculative decoding enabled, the `mamba_track_*` static buffers are never registered, so inside the graph `forward_batch.mamba_track_mask` is `None` and the mamba state-tracking write never runs. Meanwhile the scheduler still advances the ping-pong cursor (`req.mamba_next_track_idx`), so `get_mamba_ping_pong_keep_idx()` points at a slot that was never written, and `MambaRadixCache._commit_int8_checkpoint()` commits that slot. A later request matching the prefix restores a valid but wrong mamba state.

Instrumenting `_track_mamba_state_extend` on an identical workload:

| | call sites reached | track writes performed |
| --- | --- | --- |
| prefill graph **OFF** | 97,536 | **6,144** |
| prefill graph **ON** | 0 | **0** |

Instead of declining to replay when tracking is required but unsupported, the runner omits the buffers and leaves the scheduler and the radix cache believing the tracked state was written.

## Fix

Minimal correctness fix is to decline the graph for batches carrying a track mask this runner cannot honor, in `can_run_graph()`:

```python
if not self.mamba_track_enabled:
    _tm = getattr(forward_batch, "mamba_track_mask", None)
    if _tm is not None and bool(_tm.any()):
        return False
```

GSM8K n=200, `parallel=64`, three passes back-to-back with no restart, EAGLE + `extra_buffer` + `--enable-int8-mamba-checkpoint`, full 8192 bucket set:

| build | run 1 | run 2 | run 3 |
| --- | --- | --- | --- |
| unfixed | 0.965 | 0.895 | 0.890 |
| **fixed** | **0.965** | **0.955** | **0.955** |

| | track writes | prefill batches eager | prefill batches replayed |
| --- | --- | --- | --- |
| unfixed, graph ON | **0** | — | — |
| **fixed**, graph ON | **6,144** | 16 (4.3 %) | 359 |
| graph OFF (reference) | 6,144 | all | 0 |

I will make a new PR for this issue.
